daemon/algod/api/server/v2/dryrun.go

```diff
@@ -0,0 +1,512 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package v2
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/daemon/algod/api/server/v2/generated"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// DryrunRequest object uploaded to /v2/teal/dryrun
+// It is the same as generated.DryrunRequest but Txns deserialized properly.
+// Given the Transactions and simulated ledger state upload, run TEAL scripts and return debugging information.
+// This is also used for msgp-decoding
+type DryrunRequest struct {
+	// Txns is transactions to simulate
+	Txns []transactions.SignedTxn `codec:"txns"` // not supposed to be serialized
+
+	// Optional, useful for testing Application Call txns.
+	Accounts []generated.Account `codec:"accounts"`
+
+	Apps []generated.DryrunApp `codec:"apps"`
+
+	// ProtocolVersion specifies a specific version string to operate under, otherwise whatever the current protocol of the network this algod is running in.
+	ProtocolVersion string `codec:"protocol-version"`
+
+	// Round is available to some TEAL scripts. Defaults to the current round on the network this algod is attached to.
+	Round uint64 `codec:"round"`
+
+	// LatestTimestamp is available to some TEAL scripts. Defaults to the latest confirmed timestamp this algod is attached to.
+	LatestTimestamp int64 `codec:"latest-timestamp"`
+
+	Sources []generated.DryrunSource `codec:"sources"`
+}
+
+// DryrunRequestFromGenerated converts generated.DryrunRequest to DryrunRequest field by fields
+// and re-types Txns []transactions.SignedTxn
+func DryrunRequestFromGenerated(gdr *generated.DryrunRequest) (dr DryrunRequest, err error) {
+	for _, raw := range gdr.Txns {
+		// no transactions.SignedTxn in OAS, map[string]interface{} is not good enough
+		// json.RawMessage does the job
+		var txn transactions.SignedTxn
+		err = protocol.DecodeJSON(raw, &txn)
+		if err != nil {
+			return
+		}
+		dr.Txns = append(dr.Txns, txn)
+	}
+	dr.Accounts = gdr.Accounts
+	dr.Apps = gdr.Apps
+	dr.ProtocolVersion = gdr.ProtocolVersion
+	dr.Round = gdr.Round
+	dr.LatestTimestamp = int64(gdr.LatestTimestamp)
+	dr.Sources = gdr.Sources
+	return
+}
+
+func (dr *DryrunRequest) expandSources() error {
+	for i, s := range dr.Sources {
+		program, err := logic.AssembleString(s.Source)
+		if err != nil {
+			return fmt.Errorf("Dryrun Source[%d]: %v", i, err)
+		}
+		switch s.FieldName {
+		case "lsig":
+			dr.Txns[s.TxnIndex].Lsig.Logic = program
+		case "approv", "clearp":
+			for ai, app := range dr.Apps {
+				if app.AppIndex == s.AppIndex {
+					switch s.FieldName {
```

Comment:
in the case of "other", please generate an error.

Reply:
This is already handled by the outer switch on `s.FieldName`